### PR TITLE
Add a tooltip for numeric generalization in Preview step

### LINKED ISSUE #152: Show range as tooltip in cells of generalized columns
Hovering over a generalized numeric column should display the range as a tooltip. For example if the table cell shows the value `20` and is generalized with a bin size of `10`, then its hover tooltip would be `20-30`.

### PR DESCRIPTION
**Pull request by pdobacz**
Closes #152 

It's been a bit difficult to solve this, respecting all separations of concerns requirements. It came out a bit over-functional perhaps, lemme know if you can follow through.

Also, there's an issue with the *styling* of the tooltips. We have 3 options:
1. continue polishing so that it's super consistent across the app
2. be somewhat consistent
3. be noticeably inconsistent, but simplify code

So far I've chosen the (2.). Let's see if you can spot the inconsistency and/or find the code too complex and we'll see, if we need to switch to either 1 or 3. I think current choice of (2.) is "good enough".

### CODE REVIEW

File: src/AnonymizationStep/AnonymizedResultsTable.tsx [hunk 1]
```diff
@@ -21,34 +23,60 @@ type TableRowData = RowData & {
...
+    </Tooltip>
+  );
+}
+
+function cellTooltip(tooltip: string, displayValue: string) {
+  return <Tooltip title={tooltip}>{displayValue}</Tooltip>;
+}
+
 function renderValue(v: Value) {
+  // we must disable the default `ellipsis: {showTitle: true}` tooltip, so we're serving
```
Comment (cristianberneanu) [inline src/AnonymizationStep/AnonymizedResultsTable.tsx:43]: I don't understand this comment, since `showTitle` is set to `false` later on.

File: src/AnonymizationStep/AnonymizedResultsTable.tsx [hunk 2]
```diff
@@ -58,17 +86,22 @@ const mapColumn = (mode: DisplayMode) => (column: AnonymizedResultColumn, i: num
...
-          makeColumnData(column.name + ' (original)', i + '_real', AGG_COLUMN_TYPE),
+          makeColumnData(column.name + ' (original)', i + '_real', AGG_COLUMN_TYPE, renderValue),
           makeColumnData('Distortion', i + '_diff', 'real', renderRelativeNoiseValue),
         ];
     }
   }
 
-  return [makeColumnData(column.name, i.toString(), column.type)];
+  if (column.type === 'integer' || column.type === 'real') {
+    let renderTooltipValue = buildRenderTooltipValue(column.valueToTooltip);
```
Comment (cristianberneanu) [inline src/AnonymizationStep/AnonymizedResultsTable.tsx:100]: Please fix this warning.

File: src/AnonymizationStep/AnonymizedResultsTable.tsx [hunk 3]
```diff
@@ -58,17 +86,22 @@ const mapColumn = (mode: DisplayMode) => (column: AnonymizedResultColumn, i: num
...
+    return [makeColumnData(column.name, i.toString(), column.type, renderTooltipValue)];
```
Comment (cristianberneanu) [inline src/AnonymizationStep/AnonymizedResultsTable.tsx:101]: I think the call to `makeColumnData` can be extracted out.

File: src/ColumnSelectionStep/ColumnSelectionStep.tsx [hunk 4]
```diff
@@ -49,11 +51,19 @@ function minBinSize(columnType: ColumnType) {
   return columnType === 'real' ? MIN_BIN_SIZE_REAL : MIN_BIN_SIZE_INTEGER;
 }
 
+function numericGeneralizationTooltip(binSize: number, binStart: number) {
+  return binStart.toString() + '-' + (binStart + binSize).toString();
```
Comment (cristianberneanu) [inline src/ColumnSelectionStep/ColumnSelectionStep.tsx:55]: It is not clear from this that the upper bound is non-inclusive. Maybe `[0; 1)` is better?

File: src/ColumnSelectionStep/ColumnSelectionStep.tsx [hunk 5]
```diff
@@ -49,11 +51,19 @@ function minBinSize(columnType: ColumnType) {
   return columnType === 'real' ? MIN_BIN_SIZE_REAL : MIN_BIN_SIZE_INTEGER;
 }
 
+function numericGeneralizationTooltip(binSize: number, binStart: number) {
```
Comment (cristianberneanu) [inline src/ColumnSelectionStep/ColumnSelectionStep.tsx:54]: The order of the arguments is a bit weird. Usually, it is the other way around.

File: src/types/index.ts [hunk 6]
```diff
@@ -85,8 +89,11 @@ export type ResultColumn = {
 
 export type ResultRow = Value[];
 
+export type NonNullValue = boolean | number | string;
 export type Value = boolean | number | string | null;
```
Comment (cristianberneanu) [inline src/types/index.ts:93]:
```ts
export type Value = boolean | number | string;
export type NullableValue = Value | null;
```
would make more sense.

File: src/types/index.ts [hunk 7]
```diff
@@ -85,8 +89,11 @@ export type ResultColumn = {
...
 
+export type TooltipFunction = (v: NonNullValue) => string;
```
Comment (cristianberneanu) [inline src/types/index.ts:95]: Why do you need to accept `boolean | number | string` here instead of just `number`?

File: src/types/index.ts [hunk 8]
```diff
@@ -95,12 +102,18 @@ export type AnonymizedQueryResult = {
...
+      name: string;
+      type: TooltipColumnType;
+      valueToTooltip: TooltipFunction;
+    }
+  | {
+      name: string;
+      type: AnonymizedColumnType;
+    };
+
+export type AnonymizedColumnType = NoTooltipColumnType | 'aggregate';
```
Comment (cristianberneanu) [inline src/types/index.ts:116]: I don't like that tooltip information pollutes the type system.
I think it would be better to inline the types here or allow for empty `valueToTooltip` callbacks.

File: src/types/index.ts [hunk 9]
```diff
@@ -95,12 +102,18 @@ export type AnonymizedQueryResult = {
...
-  name: string;
-  type: AnonymizedColumnType;
-};
-
-export type AnonymizedColumnType = ColumnType | 'aggregate';
+export type AnonymizedResultColumn =
+  | {
+      name: string;
+      type: TooltipColumnType;
+      valueToTooltip: TooltipFunction;
```
Comment (cristianberneanu) [inline src/types/index.ts:109]: maybe `tooltipRenderer` or `tooltipGenerator`?

File: src/AnonymizationStep/AnonymizedResultsTable.tsx (hunk 1 above)
Comment (cristianberneanu) [reply src/AnonymizationStep/AnonymizedResultsTable.tsx:43]: The way the comment was phrased, I understood that we're serving tooltips everywhere in order to disable the default behavior, which is exactly the inverse of what is happening. Maybe rephrase it like this:
"Because we later disable default tooltips, we must serve a trivial `Tooltip` node everywhere."

File: src/AnonymizationStep/AnonymizedResultsTable.tsx (hunk 3 above)
Comment (cristianberneanu) [reply src/AnonymizationStep/AnonymizedResultsTable.tsx:101]: No strong opinion from me either, it can stay like this.

File: src/AnonymizationStep/AnonymizedResultsTable.tsx (hunk 1 above)
Comment (pdobacz) [reply src/AnonymizationStep/AnonymizedResultsTable.tsx:43]: yup, this is more or less the vein of the new wording. Thx!

File: src/types/index.ts (hunk 9 above)
Comment (cristianberneanu) [reply src/types/index.ts:109]: True, nothing is being rendered yet, but something is being created, made or formatted.

Comment (pdobacz) [reply src/types/index.ts:109]: I think it's down to `tooltipCallback` or `tooltipMaker` then. 

(`formatter` is too weak, since there's logic and data (`binSize`) involved.).

Actually... the difficulty in picking a name here convinces me this smells of bad abstraction. I'll spend a second to come up with an alternative to this callback model.

File: src/types/index.ts (hunk 8 above)
Comment (edongashi) [reply src/types/index.ts:116]: > I don't like that tooltip information pollutes the type system.

I agree. Can the tooltip specific types be exported from the table component file?

File: src/types/index.ts (hunk 9 above)
Comment (pdobacz) [reply src/types/index.ts:109]: see https://github.com/diffix/desktop/pull/255

**Comment by pdobacz** [comment]
Closing in favor of #255
